### PR TITLE
test(connlib): assert connection intents using property-based state machine test

### PR DESCRIPTION
**Pull request by thomaseizinger**
Opening this in a basic version that asserts sending of connection intents to resource IPs. To do this, we add some boilerplate that sets up the state machine test in general. Together with the [work](https://github.com/firezone/firezone/blob/d575dc38665b816ef7ec4a4f363d6c25b59fe56d/rust/connlib/snownet/tests/lib.rs#L296-L824) that I've done on the `snownet` tests, this can then be extended to describe the entire state machine of connlib and letting `proptest` search for inputs & combinations that break stuff.

Some more `Transition`s that I'd expect we can implement:

- Add DNS resource
- Reconnect (i.e. roam networks)
- Remove resource

The public API of `Tunnel` isn't actually very large: We add and remove resources, set upstream DNS servers and call `reconnect`. I think the bet here is that we can implement the reference state machine in a very simple way. For example, once we have added a resource and handled the connection-intent, we should be able to send an ICMP packet through the tunnel. I've already worked out how to pass `Transmit`s back and forth between relay, client and gateway (see linked `snownet` tests above). If we port that to this state machine test, we can actually exercise all the code paths that are required to encapsulate / decapsulate those packets whilst asserting against something simple like "packet pops out at the other end".

Because the setup of the test is also a proptest-strategy, we can even add the network topology as a variable by configuring the `Firewall` (see `snownet` tests) dynamically with or without blocking rules and thus force the entire tunnel through an (in-memory) relay.

Related: #4589.

### CODE REVIEW

**Comment by ReactorScram** [comment]
Might be useful. I thought draft releases didn't ping the reviewers but I guess it does 🤔

File: rust/connlib/tunnel/proptest-regressions/tests.txt
Comment (jamilbk) [inline rust/connlib/tunnel/proptest-regressions/tests.txt:1]: Reminds me of https://github.com/vcr/vcr from back in my Rails era.

**Review by jamilbk — COMMENTED** [review]
No strong feelings other than unit tests are always good

File: rust/connlib/tunnel/src/tests.rs [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+                last_connection_intents: Default::default(),
+                ip4_resource_ips: Default::default(),
+                ip6_resource_ips: Default::default(),
+            })
+            .boxed()
+    }
+
+    /// Generate a set of transitions from the current state.
+    fn transitions(state: &Self::State) -> proptest::prelude::BoxedStrategy<Self::Transition> {
+        let add_resource = cidr_resource(8).prop_map(Transition::AddCidrResource);
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/tests.rs]: We limit the size of the subnet to a host mask of 8 here, i.e. /24 for IPv4 and /120 for IPv6. Without this, iterating all IPs just takes an insane amount of CPU power for not really any benefit in terms of discovering bugs.

File: rust/connlib/tunnel/src/tests.rs [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+                state.now += Duration::from_millis(millis);
+                state.client.handle_timeout(state.now);
+                state.gateway.handle_timeout(state.now);
+            }
+        };
+
+        for (time, expected_event) in &ref_state.client_new_events {
+            assert_eq!(time, &state.now);
+            assert_eq!(expected_event, &state.client.poll_event().unwrap());
+        }
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/tests.rs:119]: So far, these are the only assertions. The properties that we test so far are:

- ICMPs to resources generate connection intents every 2s
- ICMPs to non-resources have no effect

**Comment by thomaseizinger** [comment]
@conectado This is ready for a review. Let me know if you want me to walk you through it. I've left some doc comments that explain things a bit but it took me a while to wrap my head around how this thing works so happy to explain!

File: rust/connlib/tunnel/src/tests.rs [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+                client_resources: IpNetworkTable::new(),
+                client_new_events: Default::default(),
+                last_connection_intents: Default::default(),
+                ip4_resource_ips: Default::default(),
+                ip6_resource_ips: Default::default(),
+            })
+            .boxed()
+    }
+
+    /// Generate a set of transitions from the current state.
```
Comment (conectado) [inline rust/connlib/tunnel/src/tests.rs]: I'm not too clear on this, shouldn't proptest generate the trainsitions?

File: rust/connlib/tunnel/src/tests.rs [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    fn transitions(state: &Self::State) -> proptest::prelude::BoxedStrategy<Self::Transition> {
+        let add_resource = cidr_resource(8).prop_map(Transition::AddCidrResource);
+        let tick = (0..=1000u64).prop_map(|millis| Transition::Tick { millis });
+
+        let send_unknown_icmp = (any::<IpAddr>(), any::<IpAddr>()).prop_filter_map("src != dst", {
```
Comment (conectado) [inline rust/connlib/tunnel/src/tests.rs]:
```suggestion
        let icmp_to_non_resource = (any::<IpAddr>(), any::<IpAddr>()).prop_filter_map("src != dst", {
```

maybe?

Comment (conectado) [reply rust/connlib/tunnel/src/tests.rs]: is there anything that prevents the `dst` from being a resource here?

**Review by conectado — COMMENTED** [review]
Interesting approach! I think I'm in favor of this as this can homogenize the `ClientState` tests.

Left a few questions to understand this better.

File: rust/connlib/tunnel/src/tests.rs (hunk 4 above)
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/tests.rs]: > is there anything that prevents the `dst` from being a resource here?

No, it would be by pure chance though.

File: rust/connlib/tunnel/src/tests.rs (hunk 3 above)
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/tests.rs]: It does! We describe the strategy here on how it would generate these transitions!

**Comment by thomaseizinger** [comment]
> think I'm in favor of this as this can homogenize the `ClientState` tests.

We might still need some for a few edge-cases but overall I am super optimistic to get a lot of coverage out of these!

Paired with cargo-mutants, we should be able to identify easily, what parts are still uncovered.

**Review by conectado — APPROVED** [review]
